controller/src/controller.rs

```diff
@@ -416,64 +443,41 @@ fn node_has_label(node: &Node) -> bool {
         ));
 }
 
+/// Method to verify if brupop scheduler has been set up as time window
+/// Duration between each scheduled time within time window would be 1 sec.
 #[instrument(skip())]
-fn within_time_window() -> Result<bool> {
-    let update_window_start_env = env::var(UPDATE_WINDOW_START_ENV_VAR).context(
-        controllerclient_error::MissingEnvVariableSnafu {
-            variable: UPDATE_WINDOW_START_ENV_VAR.to_string(),
-        },
-    )?;
+fn is_time_window(brupop_schedule: &Schedule) -> Result<bool> {
+    Ok(calculate_duration_between_scheduled_datetimes(brupop_schedule)?.num_seconds() == 1)
+}
 
-    let update_window_stop_env = env::var(UPDATE_WINDOW_STOP_ENV_VAR).context(
-        controllerclient_error::MissingEnvVariableSnafu {
-            variable: UPDATE_WINDOW_STOP_ENV_VAR.to_string(),
-        },
-    )?;
-
-    let current_time = Utc::now().time();
-    let update_window_start = NaiveTime::parse_from_str(&update_window_start_env, "%H:%M:%S")
-        .context(controllerclient_error::ConvertToNativeTimeSnafu {
-            date: update_window_start_env,
-        })?;
-    let update_window_stop = NaiveTime::parse_from_str(&update_window_stop_env, "%H:%M:%S")
-        .context(controllerclient_error::ConvertToNativeTimeSnafu {
-            date: update_window_stop_env,
-        })?;
-
-    // Due to the situation that controller has already started a node update but the time window will stop,
-    // we design controller to stop updating any new nodes 6 mins (the time brupop spends to update a node)
-    // before update window stop time except finishing remaining update circle. Therefore, when update window stops,
-    // no nodes are on "in-process" status.
-    let update_window_stop_with_buffer =
-        update_window_stop - chrono_duration::minutes(UPDATE_WINDOW_BUFFER);
+#[instrument(skip())]
+fn within_time_window(brupop_schedule: &Schedule) -> bool {
+    brupop_schedule.includes(Utc::now())
+}
 
-    event!(
-        Level::INFO,
-        "Calculating if current time is within update time window."
-    );
-    Ok(update_time_window_calculator(
-        &update_window_start,
-        &update_window_stop_with_buffer,
-        &current_time,
-    ))
+fn calculate_duration_between_scheduled_datetimes(
+    brupop_schedule: &Schedule,
+) -> Result<chrono::Duration> {
+    let next_schedule_datetime = brupop_schedule
+        .upcoming(Utc)
```

Comment:
Does this work when there are multiple ranges embedded in the cron, as you may have demonstrated in your test?

Reply:
Yeah,  I confirmed it on my test. `upcoming` is strong to get next scheduled datetime.  If we have something like `* * 5-9, 12-21 * * * *`, after `9:59:59` `upcoming` will recognize the next  which is `12:0:0`.